File: ch09/python/exercises.py
"""
Chapter 09: Exercises
"""

from examples import Stack


def reverse_string(string):
    stack = Stack()

    for char in string:
        stack.push(char)

    reversed_chars = []

    while stack.read():
        reversed_chars.append(stack.pop())

    return "".join(reversed_chars)


print("******** Reverse String ********")
reversed_str = reverse_string("hello")
print(f"Reversed: {reversed_str}")
assert reversed_str == "olleh"

File: ch09/go/exercises.go
package ch09

func ReverseString(str string) string {
	s := &Stack{}

	for _, char := range str {
		s.Push(char)
	}

	reversedChars := []rune{}

	for {
		if s.Read() != nil {
			reversedChars = append(reversedChars, s.Pop().(rune))
		} else {
			break
		}
	}

	return string(reversedChars)
}
